File: sdk/python/spl/parser.py
"""Tokenizer and recursive-descent parser for SPL S-expressions."""

from typing import Any

# AST type: bool | int | float | str | list[AST]
AST = Any


def tokenize(src: str) -> list[str]:
    tokens: list[str] = []
    buf = ""
    in_str = False
    for ch in src:
        if in_str:
            buf += ch
            if ch == '"':
                in_str = False
                tokens.append(buf)
                buf = ""
            continue
        if ch == '"':
            if buf.strip():
                tokens.extend(buf.strip().split())
            buf = '"'
            in_str = True
            continue
        if ch in ("(", ")"):
            if buf.strip():
                tokens.extend(buf.strip().split())
            tokens.append(ch)
            buf = ""
            continue
        buf += ch
    if buf.strip():
        tokens.extend(buf.strip().split())
    return tokens


def _atom(tok: str) -> AST:
    if tok == "#t":
        return True
    if tok == "#f":
        return False
    # Try number
    try:
        if "." in tok:
            return float(tok)
        return int(tok) if tok.lstrip("-").isdigit() else float(tok)
    except (ValueError, AttributeError):
        pass
    # Quoted string
    if tok.startswith('"') and tok.endswith('"'):
        return tok[1:-1].replace('\\"', '"')
    # Symbol (returned as string)
    return tok


def parse(src: str) -> AST:
    """Parse an SPL S-expression string into an AST."""
    tokens = tokenize(src.strip())
    if not tokens:
        raise SyntaxError("unexpected EOF")
    pos = [0]  # mutable index

    def _parse() -> AST:
        if pos[0] >= len(tokens):
            raise SyntaxError("unexpected EOF")
        tok = tokens[pos[0]]
        pos[0] += 1
        if tok == "(":
            arr: list[AST] = []
            while True:
                if pos[0] >= len(tokens):
                    raise SyntaxError("unterminated (")
                if tokens[pos[0]] == ")":
                    pos[0] += 1
                    break
                arr.append(_parse())
            return arr
        if tok == ")":
            raise SyntaxError("unexpected )")
        return _atom(tok)

    result = _parse()
    if pos[0] != len(tokens):
        raise SyntaxError("extra tokens")
    return result

File: sdk/go/spl/parser.go
package spl

import (
	"fmt"
	"strconv"
	"strings"
)

type Node interface{}

const MaxPolicyBytes = 65536 // 64 KB

func Parse(src string) (Node, error) {
	if len(src) > MaxPolicyBytes {
		return nil, fmt.Errorf("policy exceeds maximum size of %d bytes", MaxPolicyBytes)
	}
	toks := tokenize(src)
	i := 0
	var parse func() (Node, error)
	parse = func() (Node, error) {
		if i >= len(toks) {
			return nil, fmt.Errorf("unexpected EOF")
		}
		tok := toks[i]
		i++
		switch tok {
		case "(":
			var arr []Node
			for {
				if i >= len(toks) {
					return nil, fmt.Errorf("unterminated (")
				}
				if toks[i] == ")" {
					i++
					break
				}
				n, err := parse()
				if err != nil {
					return nil, err
				}
				arr = append(arr, n)
			}
			return arr, nil
		case ")":
			return nil, fmt.Errorf("unexpected )")
		default:
			if tok == "#t" {
				return true, nil
			}
			if tok == "#f" {
				return false, nil
			}
			if strings.HasPrefix(tok, "\"") && strings.HasSuffix(tok, "\"") {
				s, err := strconv.Unquote(tok)
				if err != nil {
					return nil, err
				}
				return s, nil
			}
			if n, err := strconv.ParseFloat(tok, 64); err == nil {
				return n, nil
			}
			return tok, nil
		}
	}
	return parse()
}

func tokenize(src string) []string {
	var toks []string
	var buf strings.Builder
	inStr := false
	for _, ch := range src {
		if inStr {
			buf.WriteRune(ch)
			if ch == '"' {
				inStr = false
				toks = append(toks, buf.String())
				buf.Reset()
			}
			continue
		}
		switch ch {
		case '(', ')':
			if buf.Len() > 0 {
				toks = append(toks, strings.Fields(buf.String())...)
				buf.Reset()
			}
			toks = append(toks, string(ch))
		case ' ', '\n', '\t', '\r':
			if buf.Len() > 0 {
				toks = append(toks, strings.Fields(buf.String())...)
				buf.Reset()
			}
		case '"':
			if buf.Len() > 0 {
				toks = append(toks, strings.Fields(buf.String())...)
				buf.Reset()
			}
			inStr = true
			buf.WriteRune(ch)
		default:
			buf.WriteRune(ch)
		}
	}
	if buf.Len() > 0 {
		toks = append(toks, strings.Fields(buf.String())...)
	}
	return toks
}
